### DISCUSSION TITLE
502: failed (111: Connection refused) while connecting to upstream

### DISCUSSION THREAD

**Discussion by Liege72** [mastodon/mastodon#21590] [Q&A] [answered]
All mastodon services are running without error along with nginx and redis. However, the nginx `error.log` file shows me the following:
```
2022/11/24 20:15:23 [error] 9054#9054: *1 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
```
I discovered a [mastodon help page](https://discourse.joinmastodon.org/t/502-bad-gateway-connection-refused-while-connecting-to-upstream/3396/7) with a user sharing a similar issue and I thought it would be my solution, but unfortunately not.

When trying to run `bundle exec rails assets:precompile`, I get an error saying: `ERROR: Missing RAILS_ENV environment variable, please set it to "production", "development", or "test".` This confuses me because my `puma.rb` file is configured with the `environment ENV.fetch('RAILS_ENV') { 'production' }` property.

My mastodon nginx config can be seen [here](https://hastebin.com/yecipovipo.properties).

I don't know where to go from here, does anyone have a solution or suggestion? Thanks in advance!

**Comment by afontenot** [comment]
> When trying to run `bundle exec rails assets:precompile`

The [documentation](https://docs.joinmastodon.org/admin/install/#generating-a-configuration) says to run

    RAILS_ENV=production bundle exec rake mastodon:setup

which sounds like it should solve the problem you report.

**Comment by Liege72** [comment]
I thought that would be my answer too, but unfortunately not. You can see my latest error logs below.
```
2022/11/25 19:02:53 [error] 115616#115616: *557 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:3000/", host: "23.29.118.167:443"
2022/11/25 19:07:59 [error] 115616#115616: *564 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET /.env HTTP/1.1", upstream: "http://127.0.0.1:3000/.env", host: "23.29.118.167"
2022/11/25 19:08:21 [error] 256420#256420: *1 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
2022/11/25 19:08:27 [error] 256420#256420: *3 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
```

Someone with a similar [issue](https://discourse.joinmastodon.org/t/502-bad-gateway-connection-refused-while-connecting-to-upstream/3396/6) was told to use `bundle exec rails assets:precompile` to fix their issue, but when I use this, I get the following error:
```
ERROR: Missing RAILS_ENV environment variable, please set it to "production", "development", or "test".
```

I don't know if @ClearlyClaire is around to provide some input, they seem to know a bit about this issue. I would love your help if you have some time.

Thanks in advance! :D

**Reply by Liege72** [reply]
I just checked the logs for `mastodon-streaming` and there seems to be a redis error.
```
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error! Error: connect ECONNREFUSED 127.0.0.1:6379
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1283:16)
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!  Error: connect ECONNREFUSED 127.0.0.1:6379
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1283:16) {
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!   errno: -111,
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!   code: 'ECONNREFUSED',
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!   syscall: 'connect',
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!   address: '127.0.0.1',
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error!   port: 6379
Nov 24 21:14:14 server-liege-dev node[7011]: ERR! Redis Client Error! }
```
[Discussion 18386](https://github.com/mastodon/mastodon/discussions/18386) points out a similar issue and the solution was to run `redis-server --daemonize yes`. This does not fix my issue.

**Reply by afontenot** [reply]
What happens when you run `RAILS_ENV=production bundle exec rake mastodon:setup`? It's supposed to setup Mastodon for you, including doing the asset compilation. Do you get an error? Does it appear to complete?

You can try `RAILS_ENV=production bundle exec rails assets:precompile` if you want but I don't believe you're supposed to need to do asset compilation manually like this.

Did you follow the guide in the official docs when you set this up, or something you found elsewhere?

**Reply by Liege72** [reply]
Using `RAILS_ENV=production bundle exec rake mastodon:setup` is no problem, I just run the setup and I think it's `.env.production` that gets configured. Same with `RAILS_ENV=production bundle exec rails assets:precompile`, no problem.

I followed the mastodon guide [here](https://docs.joinmastodon.org/admin/install/); pretty sure its the official guide/documentation.

I've done hours of research now and all the "solutions" I have found do not fix my issue. You specifically (afontenot), are aware of how far I've made it since my initial error. I'm convinced this is the last one I will encounter. **Hopefully!**

I do use redis for another project of mine, maybe that affects things? The connector is written in C# and I think just the default config for .NET. You can see the code below:
```cs
public static IServer server = Connector.Connection.GetServer("localhost", 6379);

public class Connector
{
    private static Lazy<ConnectionMultiplexer> lazyConnection;

    static Connector()
    {
        lazyConnection = new Lazy<ConnectionMultiplexer>(() =>
        {
            return ConnectionMultiplexer.Connect("localhost,allowAdmin=true");
        });
    }

    public static ConnectionMultiplexer Connection
    {
        get
        {
            return lazyConnection.Value;
        }
    }
}
``` 
Once again, not entirely sure if this changes anything; just trying to provide as much detail as possible.

**Reply by afontenot** [reply]
I don't know anything about redis. Hopefully someone else pops along with more knowledge. I wonder if it could be something like trying to run two instances of the server on the same port (6379). The Mastodon one could be failing because the port can't be bound, and then the connection fails to the other one (e.g. with a wrong password).

Just as a quick test you might try disabling the redis server not connected to Mastodon and then restart all the Mastodon services, and see if that helps.

A quick Internet search suggests that some people who have wanted to run multiple redis servers at once needed to switch one of them to use a different port like 6380.

**Comment by Liege72** [comment]
Sooo, I just restarted `redis-server` and now `mastodon-streaming` shows no errors? Weird because I tried restarting it many times before, but now all the sudden it works?

I checked my `nginx` logs again and it's showing the same thing as before. You can see below:
```
2022/11/26 04:48:29 [error] 106681#106681: *1 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
2022/11/26 04:51:09 [error] 108242#108242: *2 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
2022/11/26 04:51:14 [error] 108242#108242: *2 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
2022/11/26 04:51:16 [error] 108242#108242: *2 connect() failed (111: Connection refused) while connecting to upstream, client: REDACTED, server: mastodon.liege.dev, request: "GET / HTTP/2.0", upstream: "http://127.0.0.1:3000/", host: "mastodon.liege.dev"
``` 
It is no longer a redis issue, now just a `connect() failed (111: Connection refused) while connecting to upstream` issue. What that means, I do not know. I appreciate your help @afontenot and your sticking with me all this way. I understand if you are out of suggestions. Please, if anyone else has a solution, let me know. Thanks in advance.

**Reply by afontenot** [reply]
Does the status of the systemd services indicate anything interesting? E.g.

`sudo systemctl status mastodon-web`

(and also try mastodon-sidekiq and mastodon-streaming)

What about the logs? `sudo journalctl -u mastodon-web` is the primary one, but maybe try the others as well.

If the logs don't show anything, best make sure Mastodon is actually serving at `127.0.0.1:3000` as expected.

**Reply by Liege72** [reply]
The latest logs for all mastodon services are below:

**mastodon-web**
```
Nov 25 22:05:07 server-liege-dev systemd[1]: Started mastodon-web.
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] Puma starting in cluster mode...
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] * Puma version: 5.6.5 (ruby 2.7.0-p0) ("Birdie's Version")
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *  Min threads: 5
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *  Max threads: 5
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *  Environment: production
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *   Master PID: 835
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *      Workers: 2
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] *     Restarts: (✔) hot (✖) phased
Nov 25 22:05:21 server-liege-dev bundle[835]: [835] * Preloading application
Nov 25 22:05:22 server-liege-dev systemd[1]: /etc/systemd/system/mastodon-web.service:17: Unknown key name 'ProcSubset' in section 'Service', ignoring.
Nov 25 22:05:22 server-liege-dev systemd[1]: /etc/systemd/system/mastodon-web.service:18: Unknown key name 'ProtectProc' in section 'Service', ignoring.
Nov 25 22:05:56 server-liege-dev bundle[835]: [835] * Listening on http://127.0.0.1:3001
Nov 25 22:05:56 server-liege-dev bundle[835]: [835] Use Ctrl-C to stop
Nov 25 22:05:56 server-liege-dev bundle[835]: [835] - Worker 0 (PID: 1563) booted in 0.01s, phase: 0
Nov 25 22:05:56 server-liege-dev bundle[835]: [835] - Worker 1 (PID: 1566) booted in 0.01s, phase: 0
```

**mastodon-streaming**
```
Nov 26 04:45:48 server-liege-dev systemd[1]: Stopped mastodon-streaming.
Nov 26 04:45:48 server-liege-dev systemd[1]: Started mastodon-streaming.
Nov 26 04:45:50 server-liege-dev node[107312]: WARN Starting streaming API server master with 1 workers
Nov 26 04:45:51 server-liege-dev node[107336]: WARN Starting worker 1
Nov 26 04:45:51 server-liege-dev node[107336]: WARN Worker 1 now listening on 127.0.0.1:4000
```

**mastodon-sidekiq**
```
Nov 26 06:48:54 server-liege-dev bundle[106154]: 2022-11-26T06:48:54.415Z pid=106154 tid=79i class=Scheduler::AccountsStatusesCleanupScheduler jid=16d728cbe134a497e560a>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.249Z pid=106154 tid=6tq INFO: queueing Scheduler::SuspendedUserCleanupScheduler (suspended_user_cle>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.250Z pid=106154 tid=6im class=Scheduler::SuspendedUserCleanupScheduler jid=c07f958680e97669da7d867c>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.256Z pid=106154 tid=6im class=Scheduler::SuspendedUserCleanupScheduler jid=c07f958680e97669da7d867c>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.549Z pid=106154 tid=7aq INFO: queueing Scheduler::AccountsStatusesCleanupScheduler (accounts_status>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.553Z pid=106154 tid=6pe class=Scheduler::AccountsStatusesCleanupScheduler jid=a95f827110e33d5ceea23>
Nov 26 06:49:54 server-liege-dev bundle[106154]: 2022-11-26T06:49:54.563Z pid=106154 tid=6pe class=Scheduler::AccountsStatusesCleanupScheduler jid=a95f827110e33d5ceea23>
```

I also ran `lsof -nP -iTCP -sTCP:LISTEN` and found:
```
ruby2.7      835        mastodon    5u  IPv4   37314      0t0  TCP 127.0.0.1:3001 (LISTEN)
ruby2.7     1563        mastodon    5u  IPv4   37314      0t0  TCP 127.0.0.1:3001 (LISTEN)
ruby2.7     1566        mastodon    5u  IPv4   37314      0t0  TCP 127.0.0.1:3001 (LISTEN)
node      107312        mastodon   19u  IPv4 1129331      0t0  TCP 127.0.0.1:4000 (LISTEN)
```
As you can see, **mastodon is not serving port `3000`, rather `3001` and for some reason `4000` aswell (not sure if that is intended for node.** Where do I go to fix this, what file?

**Reply by afontenot** [reply]
From the [docs](https://docs.joinmastodon.org/admin/config/):

> By default, web listens on `3000` and streaming API on `4000`.

Mastodon running on 3001 instead is definitely the cause of this problem. Did you change the configuration? You can check `.env.production` for the `PORT` option; it could also be set in an environment variable but I assume you'd know if you had changed that.

Most likely, if you haven't changed anything, you will find that there's already something on your server listening on 3000, and Mastodon is automatically bumping to 3001 because it can't open that port for itself.

**Reply by Liege72** [reply]
There is nothing running on `3000` so its nothing automatic. I do recall messing with some port settings relating to `3000` and `3001`, but I don't remember the file. I do not see a `PORT` property in `.env.production`, I do not think one is added by default. What file do you think it was that I changed?

**Reply by afontenot** [reply]
Hard to say if it's not set in `.env.production`. Could be you changed the environment via the systemd unit file for mastodon-web? It kind of sounds like you've followed some suggestions you've found on some websites when looking up problems. The issue is that if they don't fix your problem they can make your configuration kind of unpredictable...

One quick way to verify that this is your problem would be to **temporarily** change the nginx config file for Mastodon.

```
upstream backend {
    server 127.0.0.1:3000 fail_timeout=0;
}
```

Change the port here to 3001 and reload nginx (*without* reloading mastodon-web!). See if it works then. If it does you should probably change it back and figure out how you managed to change the Mastodon port.

**Reply by Liege72** [reply]
It works when I change the port to `3001`. I'll look through my search history and all these files to see if I can find where I changed the port. Thank you so so much for your help, I am very exited to have my own mastodon instance running now! Curious as to what the downside is if I keep it at `3001`?

**Reply by afontenot** [reply]
The downside is that there are ways of changing the Mastodon port that aren't permanent. Some ways might go away after a reboot. Some might go away after an upgrade. Etc. It's a good idea to know where your configuration comes from and not leave it up to chance. It shouldn't hurt anything if you leave it set to that temporarily, but at the very least write down what you've done somewhere.

**Reply by Liege72** [reply]
Got it! I'll look around through my files and see where I changed it. Thank you so much for your help!

**Reply by afontenot** [reply]
No problem.

Incidentally, I missed in the docs that you actually can't define `PORT` in a .env file. So it's being defined somewhere else in your environment. I'd check your `mastodon-web` systemd unit for a `Environment` or `EnvironmentFile` declaration, that seems like the obvious place this might be set.

**Comment by Liege72** [comment, marked as answer]
### For people seeing this discussion because you are trying to fix a similar issue, please read below:

**Note:** In this case, my issue (502: bad gateway) was caused because of my changing of the default port `3000` to `3001`. This is likely NOT your issue (just the specific cause of mine) and you should verify by checking nginx and mastodon service logs. 
See mastodon logs: `journalctl -eu mastodon-*`
See nginx logs: `journalctl -eu nginx` and/or `cat /var/log/nginx/error.log`

**Context:** I had a lot of problems starting my mastodon instance and some random solution online recommended I change the default port to `3001` from `3000`. If you do this for some odd reason, make note of where you changed it because you won't want to keep this change permanently; you can see why [here](https://github.com/mastodon/mastodon/discussions/21590#discussioncomment-4241540).

As stated by @afontenot below, "If you make config changes of any kind while trying to fix a problem, make sure to write those changes down and try to revert them if they don't solve your problem. Changes to the default configuration can make debugging more difficult, doubly so if someone else is trying to help."

Hopefully this helps if you are in a similar situation!

**Reply by afontenot** [reply]
Also, some useful debugging tips for anyone else facing similar issues.

502 usually means that nginx is failing to connect to Mastodon. It's a good idea in this case to check the Mastodon logs. In this case, the Mastodon logs indicated that mastodon-web was running on the wrong port (3001 rather than 3000). Even if you don't check the docs to learn the default port, the nginx logs are enough to show that it's attempting to connect on port 3000.

403 means that nginx thinks you don't have access to the requested page. Check the nginx logs in this case. They will usually indicate a particular path (e.g. under `/home/mastodon/live/public` that the nginx user doesn't have permission to access. Make sure that the right nginx config file is being read - with a [duplicate configuration](https://github.com/mastodon/mastodon/issues/21339#issuecomment-1325632489) the settings you think you're changing might actually have no effect.

If you make config changes of any kind while trying to fix a problem, make sure to write those changes down and try to revert them if they don't solve your problem. Changes to the default configuration can make debugging more difficult, doubly so if someone else is trying to help.

**Reply by afontenot** [reply]
@Liege72 you might want to edit your answer post to not imply that the solution was changing the port to 3001. Rather, that was the *cause* of the issue.

**Comment by JRGonz** [comment]
In my case this was caused by using the instructions on the mastodon site to 'install from source' and not realizing that postgres (which is installed manually per the instruction) not running. I had to manually run the /etc/init/postgresql start script to make it function. Not sure if this helps but that's where I had my problem. I guess I have to manually create a service file that works with systemctl properly.
